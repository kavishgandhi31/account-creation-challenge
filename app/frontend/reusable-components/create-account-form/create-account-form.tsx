import React, { useState } from "react";
import { useNavigate } from "react-router-dom";
import { ErrorList } from "../error-list/error-list";
import { FormInput } from "../form-input/form-input";
import { FormButton } from "../form-button/form-button";
import { validateCredentials } from "../../utils/validate-credentials";

type CredentialsError = {
    isUsernameValid: boolean;
    isPasswordValid: boolean;
    usernameErrors: string[];
    passwordErrors: string[];
};

export const CreateAccountForm: React.FC = () => {
    const [username, setUsername] = useState<string>('');
    const [password, setPassword] = useState<string>('');
    const [credentialsError, setCredentialsError] = React.useState<CredentialsError>({
        isUsernameValid: true,
        isPasswordValid: true,
        usernameErrors: [],
        passwordErrors: []
    });
    const [fieldEntered, setFieldEntered] = useState<{ username: boolean; password: boolean }>({
        username: false,
        password: false,
    });
    const [submitting, setSubmitting] = useState(false);
    const navigate = useNavigate();

    const handleSubmit = async (e:React.FormEvent, username: string, password: string) => {
        e.preventDefault();
        setSubmitting(true);
        try {
            const { isUsernameValid, isPasswordValid, usernameErrors, passwordErrors } = await validateCredentials(username, password);
            setCredentialsError({ isUsernameValid, isPasswordValid, usernameErrors, passwordErrors });
            
            if (!isUsernameValid) {
                throw new Error(usernameErrors?.join(' ') ?? 'Invalid username');
            }
            if (!isPasswordValid) {
                throw new Error(passwordErrors?.join(' ') ?? 'Invalid password');
            }

            const response = await fetch('/api/create-account', {
                method: 'POST',
                headers: { 'Content-Type': 'application/json' },
                body: JSON.stringify({ username, password }),
            });
            const data = await response.json();

            if (response.ok) {
                navigate('/signup/account-selection');
            } else {
                throw new Error(data.errors ?? 'Failed to create account');
            }
        } catch (error) {
            console.error('Error creating account:', error);
        }
        setSubmitting(false);
    }

    const handleBlur = (event: React.FocusEvent<HTMLInputElement>) => {
        setFieldEntered((prev) => ({
            ...prev,
            [event.target.name]: true,
        }));
    };

    const handleChange = async (event: React.ChangeEvent<HTMLInputElement>) => {
        const { name, value } = event.target;
        if (name === "username") {
            setUsername(value);
            const result = await validateCredentials(value, password);
            setCredentialsError(result);
        } else if (name === "password") {
            setPassword(value);
            const result = await validateCredentials(username, value);
            setCredentialsError(result);
        }
    };
    const inputClassNames = "w-full px-3 border-b-2 border-gray-300 placeholder-gray-400 focus:outline-none focus:border-blue-500 sm:text-sm";
    const labelClassNames = "block text-sm font-small text-gray-500 mb-4";

    const isFormValid = credentialsError.isUsernameValid &&
        credentialsError.isPasswordValid &&
        username.length > 0 &&
        password.length > 0;

    return (
        <form className="w-full" onSubmit={(e) => {
            handleSubmit(e, username, password);
        }}>
            <div>
                <FormInput
                    name="username"
                    type="text"
                    label="Username"
                    required={true}
                    inputClassNames={inputClassNames} 
                    labelClassNames={labelClassNames}
                    value={username}
                    onChange={handleChange}
                    onBlur={handleBlur}
                />
                {fieldEntered.username && !credentialsError.isUsernameValid ? <ErrorList errorField="username" errors={credentialsError.usernameErrors} /> : <></>}
            </div>
            <div className="mt-4">
                <FormInput
                    name="password"
                    type="password"
                    label="Password"
                    required={true}
                    inputClassNames={inputClassNames}
                    labelClassNames={labelClassNames}
                    value={password}
                    onChange={handleChange}
                    onBlur={handleBlur}
                />
                {fieldEntered.password && !credentialsError.isPasswordValid ? <ErrorList errorField="password" errors={credentialsError.passwordErrors} /> : <></>}
            </div>
            <div className="pb-4 w-full items-center justify-center">
                <FormButton type="submit" disabled={!isFormValid || submitting} classNames={`inline-block p-4 mt-4 w-full ${isFormValid ? 'bg-[hsla(244,49%,49%,1)]' : 'bg-[hsla(244,49%,80%,0.5)]'} text-white rounded-xl`}>
                    { submitting ? "Creating account..." : "Create Account" }
                </FormButton>
            </div>
        </form>
    );
}