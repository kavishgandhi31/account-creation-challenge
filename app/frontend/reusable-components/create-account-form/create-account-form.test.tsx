import React from "react";
import '@testing-library/jest-dom';
import { render, screen, waitFor } from "@testing-library/react";
import { CreateAccountForm } from "./create-account-form";
import { BrowserRouter } from 'react-router-dom';
import { fireEvent } from "@testing-library/react";

global.fetch = jest.fn();

const mockNavigate = jest.fn();
jest.mock('react-router-dom', () => {
  const actual = jest.requireActual('react-router-dom');
  return {
    ...actual,
    useNavigate: () => mockNavigate,
  };
});

describe("Create Account Form", () => {
    beforeEach(() => {
        (global.fetch as jest.Mock).mockImplementation((url) => {
            if (url === '/users') {
                return Promise.resolve({
                    ok: true,
                    json: async () => [],
                });
            }
            if (url === '/api/create-account') {
                return Promise.resolve({
                    ok: true,
                    json: async () => ({ success: true }),
                });
            }
            return Promise.reject(new Error('Unknown endpoint'));
        });
    });

    afterEach(() => {
        jest.clearAllMocks();
    });

    it ("renders the create account form component", async() => {
        const { container } = render(<CreateAccountForm />, { wrapper: BrowserRouter });
        await waitFor(() => {
            expect(container).toMatchSnapshot();

            expect(screen.getByText("Create Account")).toBeInTheDocument();
            expect(screen.getByLabelText("Username")).toBeInTheDocument();
            expect(screen.getByLabelText("Password")).toBeInTheDocument();
            expect(screen.getByRole("button", { name: "Create Account" })).toBeInTheDocument();
        });
    });

    it ("validates that username and password fields are of the required length", async () => {
        render(<CreateAccountForm />, { wrapper: BrowserRouter });
        
        const usernameInput = screen.getByLabelText("Username") as HTMLInputElement;
        const passwordInput = screen.getByLabelText("Password") as HTMLInputElement;
        const submitButton = screen.getByRole("button", { name: "Create Account" });

        // Initial state: no warnings
        await waitFor(() => {
            expect(screen.queryByTestId("username-error")).not.toBeInTheDocument();
            expect(screen.queryByTestId("password-error")).not.toBeInTheDocument();
        });

        // Entering invalid username and password
        fireEvent.change(usernameInput, { target: { value: "ab" } });
        fireEvent.blur(usernameInput);
        fireEvent.change(passwordInput, { target: { value: "shortpass" } });
        fireEvent.blur(passwordInput);
        fireEvent.click(submitButton);
        await waitFor(() => {
            expect(screen.queryByTestId("username-error")).toHaveTextContent("Username must be between 10 and 50 characters.");
            expect(screen.queryByTestId("password-error")).toHaveTextContent("Password must be between 20 and 50 characters.");
        });
        
        // Entering valid username and password
        fireEvent.change(usernameInput, { target: { value: "validusername" } });
        fireEvent.blur(usernameInput);
        fireEvent.change(passwordInput, { target: { value: "passwordissolong123456789" } });
        fireEvent.blur(passwordInput);
        fireEvent.click(submitButton);
        await waitFor(() => {
            expect(screen.queryByTestId("username-error")).not.toBeInTheDocument();
            expect(screen.queryByTestId("password-error")).not.toBeInTheDocument();
        });
    });

    it ("validates that username field is unique, doesn't have leading/trailing spaces, and is alphanumeric", async () => {
        (global.fetch as jest.Mock).mockImplementation((url) => {
            if (url === '/users') {
                return Promise.resolve({
                    ok: true,
                    json: async () => [{ username: 'existinguser' }],
                });
            }
            return Promise.reject(new Error('Unknown endpoint'));
        });

        render(<CreateAccountForm />, { wrapper: BrowserRouter });
        
        const usernameInput = screen.getByLabelText("Username") as HTMLInputElement;
        const passwordInput = screen.getByLabelText("Password") as HTMLInputElement;
        const submitButton = screen.getByRole("button", { name: "Create Account" });

        // Entering username with leading/trailing spaces and non-alphanumeric characters
        fireEvent.change(usernameInput, { target: { value: " existing user! " } });
        fireEvent.blur(usernameInput);
        fireEvent.change(passwordInput, { target: { value: "passwordissolong123456789" } });
        fireEvent.blur(passwordInput);
        fireEvent.click(submitButton);

        await waitFor(() => {
            expect(screen.queryByTestId("username-error")).toHaveTextContent("Username can only contain letters, numbers, underscores, and hyphens");
            expect(screen.queryByTestId("username-error")).toHaveTextContent("Username cannot have leading or trailing spaces");
        });

        // Entering a duplicate username
        fireEvent.change(usernameInput, { target: { value: "existinguser" } });
        fireEvent.blur(usernameInput);
        fireEvent.click(submitButton);

         await waitFor(() => {
            expect(screen.queryByTestId("username-error")).toHaveTextContent("Username already exists. Please choose a different one.");
        });
    });

    it ("submits the form and calls the /api/create-account endpoint", async () => {
        render(<CreateAccountForm />, { wrapper: BrowserRouter });

        const usernameInput = screen.getByLabelText("Username") as HTMLInputElement;
        const passwordInput = screen.getByLabelText("Password") as HTMLInputElement;
        const submitButton = screen.getByRole("button", { name: "Create Account" });

        // Entering valid username and password
        fireEvent.change(usernameInput, { target: { value: "validusername" } });
        fireEvent.blur(usernameInput);
        fireEvent.change(passwordInput, { target: { value: "passwordissolong123456789" } });
        fireEvent.blur(passwordInput);
        fireEvent.click(submitButton);

        await waitFor(() => {
            expect((global.fetch as jest.Mock)).toHaveBeenCalledWith('/api/create-account', expect.objectContaining({
                method: 'POST',
                headers: { 'Content-Type': 'application/json' },
                body: JSON.stringify({ username: "validusername", password: "passwordissolong123456789" }),
            }));
        });

    });

    it("successful account creation navigates to account selection", async () => {
        render(<CreateAccountForm />, { wrapper: BrowserRouter });
        fireEvent.change(screen.getByLabelText(/username/i), { target: { value: "validusername" } });
        fireEvent.change(screen.getByLabelText(/password/i), { target: { value: "passwordissolong123456789" } });
        fireEvent.click(screen.getByRole("button", { name: /create account/i }));

        await waitFor(() => {
            expect(mockNavigate).toHaveBeenCalledWith("/signup/account-selection");
        });
    });
});

