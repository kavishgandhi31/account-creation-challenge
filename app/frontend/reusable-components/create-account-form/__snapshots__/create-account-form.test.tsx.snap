// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Create Account Form renders the create account form component 1`] = `
<div>
  <form
    class="w-full"
  >
    <div>
      <div>
        <label
          class="block text-sm font-small text-gray-500 mb-4"
          for="Username"
        >
          Username
        </label>
        <input
          class="w-full px-3 border-b-2 border-gray-300 placeholder-gray-400 focus:outline-none focus:border-blue-500 sm:text-sm"
          id="Username"
          name="username"
          required=""
          type="text"
          value=""
        />
      </div>
    </div>
    <div
      class="mt-4"
    >
      <div>
        <label
          class="block text-sm font-small text-gray-500 mb-4"
          for="Password"
        >
          Password
        </label>
        <input
          class="w-full px-3 border-b-2 border-gray-300 placeholder-gray-400 focus:outline-none focus:border-blue-500 sm:text-sm"
          id="Password"
          name="password"
          required=""
          type="password"
          value=""
        />
      </div>
    </div>
    <div
      class="pb-4 w-full items-center justify-center"
    >
      <button
        class="inline-block p-4 mt-4 w-full bg-[hsla(244,49%,80%,0.5)] text-white rounded-xl"
        disabled=""
        type="submit"
      >
        Create Account
      </button>
    </div>
  </form>
</div>
`;
